### PR TITLE
[spark] Retry data evolution compaction after merge conflicts

### PR DESCRIPTION
**Pull request by JingsongLi**
## What changed

- introduce a typed data-evolution row-range conflict for retry eligibility
- retry Spark data-evolution compaction only for that conflict type
- rebase compatible concurrent partial-column files onto the staged compact row-id boundaries
- validate concurrent files by their portable file contract instead of `Snapshot.operation`

## Why

Data-evolution compaction can race with `MERGE INTO`: compaction changes row-id range boundaries while MERGE writes ordinary partial-column files against the previous boundaries. The commit then fails even though the concurrent changes can be preserved by rewriting only those partial columns onto the new compact ranges.

Python MERGE does not persist `Snapshot.operation`, so operation-based detection would miss valid conflicts. This change recognizes the safe file shape directly while retaining the typed row-range conflict as the only retry trigger. Other conflict types still fail immediately.

## Impact

Spark compaction can recover from compatible concurrent Spark or Python MERGE updates without rerunning the full compaction. The retry reads the latest snapshot, includes the concurrent partial-column files, and rewrites only the affected fields and row-id ranges.

## Validation

- targeted `ConflictDetectionTest` row-range conflict tests
- Spark 3.5 `CompactProcedureTest`: `rebase data evolution compact after operation-less partial update`
- `mvn -pl paimon-spark/paimon-spark-common -Pspark3 -DskipTests compile`
- `git diff --check`

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+          val updatedFields = table
+            .rowType()
+            .getFieldNames
+            .asScala
+            .filter(name => files.exists(_.file.writeCols().contains(name)))
+            .toSeq
+          if (updatedFields.isEmpty) {
+            return JOptional.empty()
+          }
+          Some(TargetRewrite(target, files, updatedFields))
```
Comment (leaves12138) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala]: This does not compile with Scala 2.13 (and therefore breaks the Spark 4 build). With `scala.collection.JavaConverters`, the converted/grouped values are mutable `Buffer`s, while the method and case-class parameters use Scala 2.13's immutable `Seq`. There are at least two mismatches: `files` here and `rewrites` passed to `rewriteFiles` below. Please convert both explicitly (for example, `files.toSeq` and `rewrites.toSeq`) or type these APIs as `scala.collection.Seq`.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+      target =>
+        val files = overlappingAdditions.filter(target.contains)
+        if (files.nonEmpty && files.exists(file => file.file.nonNullRowIdRange() != target.range)) {
+          val updatedFields = table
+            .rowType()
+            .getFieldNames
+            .asScala
+            .filter(name => files.exists(_.file.writeCols().contains(name)))
+            .toSeq
+          if (updatedFields.isEmpty) {
```
Comment (leaves12138) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala:130]: This can silently lose updates when the schema evolves concurrently. If compaction starts with an old schema and a concurrent partial update writes both an existing column and a newly added column, `updatedFields` is filtered through the stale `table.rowType()`, so the new column is omitted. The rewritten compact increment then deletes the original partial file but replaces it with a file containing only the old column. Since the commit uses the latest schema, it succeeds while the new-column update is lost.

I reproduced this with an old `(id, value)` compaction, concurrent `ADD COLUMN extra`, and a partial update of `(value, extra)`: before the rebase the row is `(1, 11, 99)`, but after the compact commit it becomes `(1, 11, NULL)`.

Please either reject rebasing when the schema changed between the preparation and latest snapshots, or rebuild the rewrite against the latest schema while preserving every column in the concurrent partial file.

**Review by leaves12138 — APPROVED** [review]
Re-reviewed the latest changes. The schema-evolution race now rejects the compact rebase and preserves the concurrent update, while the compatible partial-update rebase still succeeds. The targeted Spark tests and ConflictDetectionTest pass.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    val currentSplits = targetReader(latestSnapshot, targetRewrites.map(_.target))
+      .read()
+      .splits()
+      .asScala
+      .collect { case split: DataSplit => split }
+      .toSeq
+
+    val rewrittenMessages = targetRewrites
```
Comment (leaves12138) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala:149]: `rewriteFiles` stages real data files, but this `flatMap` can invoke it multiple times for different `updatedFields` groups. If one group succeeds and a later group fails (for example, a Spark task/write failure), `rewrite` throws before returning any commit messages, so `commitWithMergeConflictRetry` never receives the already-created messages and cannot abort them. Those files are left as orphans even though no commit was attempted and cleanup is still unambiguously safe. Could we accumulate the staged messages in a `try` block and abort all previously produced messages on any subsequent rewrite failure (or make the writer expose an abortable rewrite result)? A failure-injection test with two field groups would also protect this path.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+  private def normalRowIdFiles(files: Iterable[DataFileMeta]): Seq[DataFileMeta] = {
+    files.filter(isNormalRowIdFile).toSeq
+  }
+
+  private def isNormalRowIdFile(file: DataFileMeta): Boolean = {
+    file.firstRowId() != null && !isBlobFile(file.fileName()) && !isVectorStoreFile(file.fileName())
+  }
+
+  private def isRegularPartialFile(file: DataFileMeta): Boolean = {
```
Comment (leaves12138) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/DataEvolutionCompactMergeConflictRewriter.scala:420]: This shape also accepts normal files produced by another data-evolution COMPACT. `DataEvolutionNormalCompactTask` writes `FileSource.COMPACT`, but on a table with dedicated BLOB/vector fields its normal output has a non-empty user-column `writeCols`, so it passes this predicate.

That misclassification creates a data-loss race. Suppose compact B is staged, compact A commits the same new row-range boundary, and B rebases A's COMPACT output here. The rebased partial file is stamped with B's later snapshot sequence. If a MERGE updates the same column after this rewrite finishes but before B commits, its file uses A's already-current boundary, so all ranges are identical and the COMPACT conflict check does not retry. B's older rebased values then have the newer sequence and override that MERGE update.

Please exclude `FileSource.COMPACT` (the portable MERGE partial files are APPEND files) and add a race test where a same-boundary update lands between rewrite and commit.

**Review by leaves12138 — CHANGES_REQUESTED** [review]
I found two production-blocking issues in the retry path: partial rewrite failures can leak already-staged retry files, and normal output from a concurrent data-evolution COMPACT can be misclassified as a MERGE partial file, creating a same-boundary update loss race. Please address the inline comments before merging.

**Review by leaves12138 — APPROVED** [review]
Re-reviewed the latest head in detail. The concurrent data-evolution COMPACT output is now explicitly marked as COMPACT and is excluded from MERGE partial-file rebasing. The rebased bridge files also preserve the source sequence range, so a MERGE committed after the rewrite remains newer. I verified the operation-less Python-compatible path, schema-evolution rejection, concurrent compaction before and after rewrite, bounded planner batches, retry cleanup, and sequence ordering. ConflictDetectionTest (45 tests) and Spark 3.5 CompactProcedureTest (45 tests, including all 8 new concurrency/retry cases) pass locally. The accepted orphan-file risk is non-blocking.
